odoo/domains.py

```diff
@@ -0,0 +1,1484 @@
+            parent_model = model.env[field.related_field.model_name]
+            parent_fname = model._inherits[parent_model._name]
+            parent_domain = self.optimize(parent_model)
+            return DomainLeaf(parent_fname, 'any', parent_domain)
+
+        # handle non-stored fields (replace by queryable/stored items)
+        if not field.store:
+            if property_name:
+                # just hope _condition_to_sql handles this expression
+                return self
+            # find the implementation of search and execute it
+            if not field.search:
+                _logger.error("Non-stored field %s cannot be searched.", field)
+                if _logger.isEnabledFor(logging.DEBUG):
+                    _logger.debug(''.join(traceback.format_stack()))
+                return _TRUE_DOMAIN
+            operator, value = self.operator, self.value
+            if operator in ('in', 'not in') and len(value) == 1:
+                # a lot of implementations expect '=' or '!=' operators
+                operator = '=' if operator == 'in' else '!='
+                value = next(iter(value))
+            computed_domain = field.determine_domain(model, operator, value)
+            return Domain(computed_domain).optimize(model)
+
+        # optimizations based on operator and on field type
+        leaf = self
+        for opt in _LEAF_OPTIMIZATIONS_BY_OPERATOR[leaf.operator]:
+            leaf = opt(leaf, model)
+            if leaf is not self:
+                return leaf.optimize(model)
+        for opt in _LEAF_OPTIMIZATIONS_BY_FIELD_TYPE[field.type]:
+            leaf = opt(leaf, model)
+            if leaf is not self:
+                return leaf.optimize(model)
+
+        # asserts after optimization
+        operator = self.operator
+        if operator not in STANDARD_TERM_OPERATORS:
+            self._raise("Not standard operator left")
+
+        if (
+            not field.relational
+            and operator in ('any', 'not any')
+            and field.name != 'id'  # can use 'id'
+            # # Odoo internals use ('xxx', 'any', Query), check only Domain
+            and (not field.store or isinstance(self.value, Domain))
+        ):
+            self._raise("Cannot use 'any' with non-relational fields")
+
+        return self
+
+    def _to_sql(self, model: BaseModel, alias: str, query: Query) -> SQL:
+        field, operator, value = self.field, self.operator, self.value
+        if operator in ('in', 'not in') and isinstance(value, COLLECTION_TYPES) and len(value) == 1:
+            # use "=" for backwards compatbility when building the condition
+            operator = '=' if operator == 'in' else '!='
+            value = next(iter(value))
+        return model._condition_to_sql(alias, field, operator, value, query)
+
+
+# --------------------------------------------------
+# Optimizations
+# --------------------------------------------------
+
+_NARY_OPTIMIZATIONS = list()
```

Comment:
TODO Should we merge optimizations into the DomainNary, to DomainLeaf and to Field?